src/nccl_ofi_net.c

```diff
@@ -25,13 +25,68 @@ nccl_ofi_t **nccl_ofi_component = NULL;
 bool local_mr = false;
 /* Indicates if memory registration of device buffers is required */
 bool hmem_mr = false;
+/* Indicates if endpoint memory registration is required */
+bool endpoint_mr = false;
+/* Indicates if remote virtual addressing is used */
+bool virt_addr_mr = false;
+/* Indicates if the provider selects MR keys */
+bool prov_key_mr = false;
 /* Indicates if GPUDirect is supported by libfabric provider */
 bool support_gdr = true;
 /* Indicates if the cudaDeviceFlushGPUDirectRDMAWrites function should be used
  * to flush data to the GPU. Note, CUDA flush support is not supported on all
  * platforms and should be disabled by default */
 bool cuda_flush = false;
 
+/* Table indicating allocation state of MR keys */
+static bool *mr_keys = NULL;
```

Comment:
Yeah, this struck me as odd in how the code is structured. There's no good place for us to free memory that's allocated in the `ofi_init` function (e.g. `nccl_ofi_component`). A good change to make here would be to have per device MR key tables in `nccl_ofi_component`. This is primarily an optimization so that we use keys more efficiently and is not immediately necessary. I can work on this change as a follow-up PR to avoid a messy refactor that will complicate code review and lengthen the merge process. Will be much easier/faster to review as a separate change.

Reply:
Scratch the above -- the MR key table already tracks devices separately, so there's not much to do here.